src/transformers/models/clvp/modeling_clvp.py

```diff
@@ -0,0 +1,1873 @@
+        # This logic is only used for the attention in CLVPConditioningEncoder. For the attention of AutoRegressive,
+        # speech and text models it is not used.
+        if self.apply_hidden_states_norm:
+            hidden_states = torch.permute(self.norm(torch.permute(hidden_states, (0, 2, 1))), (0, 2, 1))
+
+        bsz, tgt_len, embed_dim = hidden_states.size()
+
+        # get query proj
+        query_states = self.q_proj(hidden_states) * self.scale
+        key_states = self._shape(self.k_proj(hidden_states), -1, bsz)
+        value_states = self._shape(self.v_proj(hidden_states), -1, bsz)
+
+        if layer_past is not None:
+            past_key, past_value = layer_past
+            key_states = torch.cat((past_key, key_states), dim=-2)
+            value_states = torch.cat((past_value, value_states), dim=-2)
+
+        if use_cache is True:
+            present = (key_states, value_states)
+        else:
+            present = None
+
+        proj_shape = (bsz * self.num_heads, -1, self.head_dim)
+        query_states = self._shape(query_states, tgt_len, bsz).view(*proj_shape)
+        key_states = key_states.view(*proj_shape)
+        value_states = value_states.view(*proj_shape)
+
+        if rotary_pos_emb is not None:
+            query_states = apply_rotary_pos_emb(query_states, rotary_pos_emb)
+            key_states = apply_rotary_pos_emb(key_states, rotary_pos_emb)
+            value_states = apply_rotary_pos_emb(value_states, rotary_pos_emb)
+
+        src_len = key_states.size(1)
+        attn_weights = torch.bmm(query_states, key_states.transpose(1, 2))
+
+        if attn_weights.size() != (bsz * self.num_heads, tgt_len, src_len):
+            raise ValueError(
+                f"Attention weights should be of size {(bsz * self.num_heads, tgt_len, src_len)}, but is"
+                f" {attn_weights.size()}"
+            )
+
+        # apply the causal_attention_mask first
+        if use_causal_attention_mask:
+            # if causal mask if given then apply it directly
+            if causal_attention_mask is not None:
+                attn_weights = attn_weights.view(bsz, self.num_heads, tgt_len, src_len) + causal_attention_mask
+                attn_weights = attn_weights.view(bsz * self.num_heads, tgt_len, src_len)
+
+            # if there is no causal mask given but the config has `max_position_embeddings`, then use the constructed
+            # mask. This portion mimics the `GPT2Attention`.
+            elif causal_attention_mask is None and hasattr(self.config, "max_position_embeddings"):
+                query_length, key_length = query_states.size(-2), key_states.size(-2)
+                causal_mask = self.bias[:, :, key_length - query_length : key_length, :key_length]
+                mask_value = torch.finfo(attn_weights.dtype).min
+                mask_value = torch.full([], mask_value, dtype=attn_weights.dtype).to(attn_weights.device)
```

Comment:
pretty sure the mask should be created outside the `attention` layer for less overhead

Reply:
This is to mimic the `GPT2Attention` part,  as you can see [here](https://github.com/huggingface/transformers/blob/0a55d9f7376f72ad3ff296d4249840021b03bcc4/src/transformers/models/gpt2/modeling_gpt2.py#L194). If I change something here, I might need to change the `ClvpDecoderBlock`(which is just a `GPT2Block`) thus need to remove the `# Copied from` statements from `ClvpDecoderBlock.forward`.

Should I proceed creating the mask outside the `attention` layer or keep it same?